### PR TITLE
First pass at generics refactoring

### PR DESCRIPTION
**Pull request by thebrianchen**
I introduced generics into CollectionType and annotated the the public method return types on. Can you make sure that I was correct?

A couple of other questions
- Should I refactor use cases of `primitiveType()` to call `getContainedType()` instead (eventually after the generics have been introduced into other Types)?
- What other use cases can I look for to update now that generics are in CollectionType? I tried searching for "type as" and "type() as" but couldn't find other ones.

### CODE REVIEW

File: src/runtime/type.ts [hunk 1]
```diff
@@ -381,42 +381,42 @@ export class TypeVariable extends Type {
 }
 
 
-export class CollectionType extends Type {
-  readonly collectionType: Type;
+export class CollectionType<T extends Type> extends Type {
+  readonly collectionType: T;
 
-  constructor(collectionType: Type) {
+  constructor(collectionType: T) {
```
Comment (lindner) [inline src/runtime/type.ts]: can this also be `T extends Type` ?

File: src/runtime/type.ts [hunk 2]
```diff
@@ -381,42 +381,42 @@ export class TypeVariable extends Type {
...
     this.collectionType = collectionType;
   }
 
-  get isCollection() {
+  get isCollection(): boolean {
     return true;
   }
 
-  mergeTypeVariablesByName(variableMap: Map<string, Type>) {
+  mergeTypeVariablesByName(variableMap: Map<string, Type>): CollectionType<Type> {
```
Comment (lindner) [inline src/runtime/type.ts]: CollectionType<T>?

File: src/runtime/type.ts (hunk 1 above)
Comment (thebrianchen) [reply src/runtime/type.ts]: Adding that does not compile. I believe that the `extends Type` is inferred by the TS compiler.

File: src/runtime/type.ts (hunk 2 above)
Comment (thebrianchen) [reply src/runtime/type.ts]: In this case, `const result` is of type `Type`, so `result.collectionOf()` has type `Collection<Type>`.

So the best we can do here is to constrain the return type to `CollectionType<Type>`.

File: src/runtime/test/interface-info-test.ts [hunk 3]
```diff
@@ -202,7 +202,7 @@ describe('interface', () => {
     // Initially handle type are unresolvable type variables.
     assert.lengthOf(recipe.handles, 2);
     for (const handle of recipe.handles) {
-      const collectionType = (handle.type as CollectionType).collectionType;
+      const collectionType = (handle.type as CollectionType<Type>).collectionType;
```
Comment (lindner) [inline src/runtime/test/interface-info-test.ts]: Can you try just removing the 'as CollectionType<Type>' syntax?  The generics should be working in this case.

Comment (thebrianchen) [reply src/runtime/test/interface-info-test.ts]: VScode doesn't seem to like it and marks `handle.type` as type `Type`. From my budding understanding of generics, the cast to `CollectionType<Type>` is necessary b/c `handle.type` could be of any `Type`. 

However, once it is cast to `Collection<Type>` and we obtain the underlying `collectionType`, the `collectionType` no longer needs to be cast. For example, in the case of `test/recipes/particle-tests.ts`, once `listConnType` is cast to a `Collection<TypeVariable>`, `listConnType.collectionType` is automatically assumed to be of type `TypeVariable` and would no longer need to be cast thanks to the generics.

File: src/runtime/type.ts [hunk 4]
```diff
@@ -134,7 +134,7 @@ export abstract class Type {
   }
 
   collectionOf() {
-    return new CollectionType(this);
+    return new CollectionType<Type>(this);
```
Comment (lindner) [inline src/runtime/type.ts]: The type will vary based on the `this` value.  So you should not include the explicit <Type> here.

**Comment by thebrianchen** [comment]
Context: #2561

File: src/runtime/test/recipe/particle-tests.ts [hunk 5]
```diff
@@ -37,8 +37,8 @@ describe('Recipe Particle', () => {
       const type = hostedParticleConn.type as InterfaceType;
       const ifaceVariable = type.interfaceInfo.handles[0].type as TypeVariable;
 
-      const listConnType = listConn.type as CollectionType;
-      const listUnpackedVariable = listConnType.collectionType as TypeVariable;
+      const listConnType = listConn.type as CollectionType<TypeVariable>;
+      const listUnpackedVariable = listConnType.collectionType; // *question* as opposed to getContainedType()?
```
Comment (lindner) [inline src/runtime/test/recipe/particle-tests.ts]: remove comment

File: src/runtime/test/type-checker-tests.ts [hunk 6]
```diff
@@ -101,8 +101,8 @@ describe('TypeChecker', () => {
     const a = TypeVariable.make('a').collectionOf();
     const b = TypeVariable.make('b').collectionOf();
     const resolution = EntityType.make(['Thing'], {});
-    (a.collectionType as TypeVariable).variable.resolution = resolution;
-    (b.collectionType as TypeVariable).variable.resolution = resolution;
+    (a.collectionType).variable.resolution = resolution;
```
Comment (lindner) [inline src/runtime/test/type-checker-tests.ts]: yay!  this is where generics start to help us.

Comment (lindner) [reply src/runtime/test/type-checker-tests.ts]: also, remove parens

**Comment by lindner** [comment]
Shane, would you like to do final signoff on this?

**Pull request merged**